pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -521,18 +529,43 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
 //  fqdn/NameManager instance).
 //  - Write the response to the endpoint.
 func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
-	stat := ProxyRequestContext{}
-	stat.ProcessingTime.Start()
 	requestID := request.Id // keep the original request ID
 	qname := string(request.Question[0].Name)
 	protocol := w.LocalAddr().Network()
+	epIPPort := w.RemoteAddr().String()
 	scopedLog := log.WithFields(logrus.Fields{
 		logfields.DNSName:      qname,
-		logfields.IPAddr:       w.RemoteAddr(),
-		logfields.DNSRequestID: request.Id})
+		logfields.IPAddr:       epIPPort,
+		logfields.DNSRequestID: requestID,
+	})
+
+	var stat ProxyRequestContext
+	if p.ConcurrencyLimit != nil {
+		// TODO: Consider plumbing the daemon context here.
+		ctx, cancel := context.WithTimeout(context.TODO(), 1*time.Second)
+		defer cancel()
+
+		stat.SemaphoreAcquireTime.Start()
+		if err := p.ConcurrencyLimit.Acquire(ctx, 1); err != nil && errors.Is(err, context.DeadlineExceeded) {
+			stat.SemaphoreAcquireTime.End(false)
+			err = fmt.Errorf(
+				"failed to acquire DNS proxy semaphore, %d parallel requests already in-flight: %w",
+				option.Config.DNSProxyConcurrencyLimit,
+				err,
+			)
+			scopedLog.WithError(err).Error("Dropping DNS request due to too many DNS requests already in-flight")
```

Comment:
This one should probably get ratelimited at some point, since this is triggered based on incoming traffic.

Reply:
Good point. Maybe something we can do in a followup cc @nebril

Reply:
Added